src/dwarf/resolver.rs

```diff
@@ -70,16 +70,26 @@ where
 
 /// A type managing lookup of symbols.
 struct DebugSyms<'mmap> {
-    /// Debug symbols, ordered by name.
+    /// Debug symbols, ordered by address.
     syms: Box<[DWSymInfo<'mmap>]>,
+    /// An index on top of `syms` sorted by name.
+    // TODO: The index could be optimized to use smaller-than-word-size integers
+    //       if we work with only a few symbols.
+    by_name_idx: Box<[usize]>,
 }
 
 impl<'mmap> DebugSyms<'mmap> {
     /// Create a new `DebugSyms` object from the given set of symbols.
-    fn new(mut syms: Vec<DWSymInfo<'mmap>>) -> Self {
-        let () = syms.sort_by_key(|sym| sym.name);
+    fn new(syms: Vec<DWSymInfo<'mmap>>) -> Self {
```

Comment:
do we use DebugSyms both for looking up functions (STT_FUNCs) and variables (STT_OBJECTs)? If not, it's probably a good idea to filter down to just specific type of symbols?

And in any case, there are tons of other types (e.g., STT_TLS and STT_FILE) that are irrelevant, so best get rid of them early on to not waste memory on them.

Note that for STT_FUNC we might want to keep also STT_GNU_IFUNC...

Reply:
Isn't `STT_FUNC` an ELF-ism? `DebugSyms` is all about DWARF and DWARF only. Currently it only collects and reports functions.

Reply:
My bad, probably wrong place to comment on. My question was triggered by ElfResolver's `find_syms` implementation, where we have `parser.find_sym(addr, STT_FUNC)` call. So I was just wondering if we have to do any additional filtering due to STT_FUNC instead of just making sure whatever cache we have already contains just functions.

And then also this whole STT_GNU_IFUNC thing which would be ignored right now, but I think does come up in stack traces.

Reply:
Yeah, `ElfResolver` currently just takes everything in `.symtab` and works with that. The section is memory mapped and we just store references to said memory in a vector, which is sorted. So it's ~8 bytes per symbol. But yes, we could and probably should shrink that further by removing stuff we don't need.

As for `STT_GNU_IFUNC`, yes, we probably should support it. I have opened https://github.com/libbpf/blazesym/issues/210.